### PR TITLE
Add confusion matrix and precision-recall curve support (#132)

### LINKED ISSUE #132: Confusion matrix, Precision-recall curve, F1 curve; Precision, recall and confidence curve
Hello dear. Is it possible to integrate the Confusion matrix, Precision-recall curve, F1 curve; Precision, recall and confidence curve.
For research purposes, I am currently blocked at the task. Thank you for your good work

### PR DESCRIPTION
**Pull request by Moon-Wrecker**
## Description

Adds confusion matrix and precision-recall curve support for training and evaluation. Metrics can be saved as PNG, CSV, and JSON files with configurable confidence and IoU thresholds.

Includes tests and README updates.

Closes #132

___

## **CodeAnt-AI Description**
Add configurable detection metrics to evaluation and training

### What Changed
- Evaluation saves regular and normalized confusion matrices, precision-recall curves, CSV data, and a JSON summary
- Training can save the same metrics after each validation epoch
- Detection confidence and box-matching thresholds can be configured, with validation for invalid values
- Evaluation results use new numbered directories or a user-selected output directory, avoiding existing result files
- Reports handle sparse classes, missed and duplicate detections, crowd annotations, empty inputs, and distributed evaluation
- Added tests and usage guidance for metric generation and edge cases

### Impact
`✅ Clearer per-class detection errors`
`✅ Exportable evaluation results`
`✅ Configurable training and evaluation thresholds`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -223,8 +223,30 @@ def parse_opt():
...
+        default=0.25,
+        type=float,
+        help='detection threshold for the confusion matrix'
+    )
+    parser.add_argument(
+        '--iou-threshold',
+        dest='iou_threshold',
+        default=0.50,
+        type=float,
+        choices=[0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95],
```
[the remark this answers is not in the document]
Comment (Moon-Wrecker) [reply train.py:243]: Thanks for flagging this. The restriction is intentional because --iou-threshold selects both the confusion-matrix matching threshold and a precision-recall slice from the COCO evaluation output. COCO/TorchMetrics provides the standard 0.50 to 0.95 grid in 0.05 steps, so 0.51 has no corresponding PR slice. Supporting arbitrary values would require changing the evaluator’s IoU grid and could alter the existing COCO mAP behavior, which this PR preserves.

File: eval.py [hunk 2]
```diff
@@ -100,12 +138,14 @@
...
     create_model = create_model[args['model']]
     if args['weights'] is None:
+        coco_model = False
         try:
             model, coco_model = create_model(num_classes=NUM_CLASSES, coco_model=True)
         except:
             model = create_model(num_classes=NUM_CLASSES, coco_model=True)
         if coco_model:
             COCO_91_CLASSES = data_configs['COCO_91_CLASSES']
+            EVAL_CLASSES = COCO_91_CLASSES
```
[the remark this answers is not in the document]
Comment (Moon-Wrecker) [reply eval.py:148]: The 91-position COCO class list must still be retained because pretrained COCO models use sparse category IDs up to 90. However, the unused N/A entries do not need to appear in the confusion matrix. In f87c409, the matrix now maps active category IDs to compact positions while preserving the correct COCO label mapping and omitting unused N/A axes. A regression test for sparse category IDs was also added.

File: eval.py [hunk 3]
```diff
@@ -139,7 +183,16 @@ def evaluate(
         classes=None,
         colors=None
     ):
-        metric = MeanAveragePrecision(class_metrics=args['verbose'])
+        metric = MeanAveragePrecision(
+            class_metrics=args['verbose'],
+            extended_summary=args['save_metrics']
```
[the remark this answers is not in the document]
Comment (Moon-Wrecker) [reply eval.py:188]: `extended_summary` is enabled only when metric artifacts are requested because its precision output is required for the PR curve. Standalone evaluation saves metrics by default, while `--no-save-metrics` disables both artifact generation and the extended summary; training metrics remain opt-in. The extended values are consumed once and removed before printing, allowing the PR curve to reuse the existing evaluation instead of requiring another inference pass.

File: utils/general.py [hunk 4]
```diff
@@ -309,6 +504,25 @@ def set_infer_dir():
...
+    """
+    if project_dir != None:
+        os.makedirs(project_dir, exist_ok=True)
+        return project_dir
+    if not os.path.exists('outputs/evaluation'):
+        os.makedirs('outputs/evaluation')
+    num_eval_dirs_present = len(os.listdir('outputs/evaluation/'))
+    next_dir_num = num_eval_dirs_present + 1
+    new_dir_name = f"outputs/evaluation/res_{next_dir_num}"
+    os.makedirs(new_dir_name, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (Moon-Wrecker) [reply utils/general.py]: This initially followed the directory-numbering convention already used by the training and inference helpers, but concurrent evaluations could still select the same directory. In f87c409, candidate directories are now created atomically, with the function retrying the next res_N value if the directory already exists. A regression test covering existing directory collisions was also added.

File: eval.py [hunk 5]
```diff
@@ -180,39 +243,79 @@ def evaluate(
         torch.set_num_threads(n_threads)
         metric.update(preds, target)
         metric_summary = metric.compute()
+        if args['save_metrics']:
+            PR_data = get_precision_recall(
+                metric_summary['precision'],
+                classes,
+                metric_summary['classes'],
```
[the remark this answers is not in the document]
Comment (Moon-Wrecker) [reply eval.py:250]: Thanks for flagging this. I verified the exact configuration with TorchMetrics 1.9.0 using `class_metrics=False` and `extended_summary=True`. The computed result still includes the `classes` key. The `class_metrics` option only controls the calculation of `map_per_class` and `mar_100_per_class`; the observed class IDs remain part of the result and are used here to resolve the precision-recall curve labels. Therefore, `metric_summary['classes']` does not raise a `KeyError`, and no code change is required.

**Comment by Moon-Wrecker** [comment]
Hi @sovit-123 , Please review this PR.
